doc/manual/man/man1/init_exp.1
.TH "init_exp" 1 "" "" "Staden Package"
.SH "NAME"
.PP
init_exp \- create and initialise an Experiment File

.SH "SYNOPSIS"
.PP

\fBinit_exp\fP [\fB-\fP(\fBabi\fP|\fBalf\fP|\fBscf\fP|\fBpln\fP)]
[\fB-output\fP \fIfile\fP] [\fB-name\fP \fIentry_name\fP] [\fB-conf\fP] \fIfile\fP

.SH "DESCRIPTION"
.PP

\fBinit_exp\fP initiates an Experiment File for a binary trace file or
a plain sequence file. The Experiment File created
contains the \fBID\fP, \fBEN\fP, \fBLN\fP, \fBLT\fP and \fBSQ\fP
lines.

The experiment file is, by default, sent to standard output, unless an
output file is specified using the \fB-output\fP option. The default
entry name for the Experiment File is derived from the filenames used.
If an output file has been specified, then this is taken as the
\fBEN\fP field. Otherwise the input file name is used. The user can
override the default by using the \fB-name\fP option.

.SH "OPTIONS"
.PP

.TP
\fB-abi\fP, \fB-alf\fP, \fB-scf\fP, \fB-pln\fP
Specify an input file format. This is not usually required as
\fBinit_exp\fP will automatically determine the correct input file
type. This option is supplied incase the automatic determination is
incorrect (which is possible, but has never been observed).

.TP
\fB-output\fP \fIfile\fP
The experiment file will be written to \fIfile\fP instead of standard
output. Additionally the value of the \fBEN\fP and \fBID\fP
fields, assuming \fB-name\fP has not been specified, will be \fIfile\fP.

.TP
\fB-name\fP \fIname\fP
Sets the \fBID\fP and \fBEN\fP fields to \fIname\fP, regardless of
the output filename used.

.TP
\fB-conf\fP
Fills out the \fBAV\fP field with the quality values found in the SCF
file.

.TE
.SH "NOTES"
.PP

This program was formerly known as \fBexpGetSeq\fP.

.SH "SEE ALSO"
.PP

_fxref(Formats-Exp, ExperimentFile(4), formats)
\fBRead\fP(4)


NEWS
-----------------------------------------------------------------------------
30th September 2005
===================

Version 1.6.0 tagged

-----------------------------------------------------------------------------

27th September 2005
===================

Gap4
----

* Renamed the discrepancy plot to be second highest confidence plot.

* Fixed some (apparently rare) find internal joins crashes.

* Added a minimum quality slider to the Report Mutations "by
  differences" mode. Also fixed crashes and missing pictures in the
  HTML report.

io_lib
------

* Changed the HashFile format and updated it to be official version
  1.00.

  The key difference is that there's now a footer pointing back into
  the hashfile header allowing for the index to simply be appended to
  a (e.g. tar) archive.

* Improved handling of 454 data in ZTR; it no longer attempts to
  compress raw pyrosequencing data as this was not working well.

* Fixed Bug ID 1289095 (file descriptor leak).

-----------------------------------------------------------------------------

24th August 2005
================

Gap4
----

* Rewrote the Shuffle Pads algorithm to be more like ReAligner. It can
  introduce new pads and remove old ones as necessary in order to
  optimise the alignments. It has also moved from inside the contig
  editor to part of the main menu so that it can be scripted easier.

* The join editor invoked from Find Read Pairs plot now brings up the
  editors at the start/end of the sequences forming the read-pair
  instead of the start/end of the two contigs.

* Search by sequence in the editor now finds consensus matches
  too. Also fixed some bugs where cutoff data matches were not always
  found and in some cases (reverse search) non-cutoff matches could be
  missed.

* The SNP Candidates plot now allows for the consensus to be saved
  directly from this window. The consensus and fake sequences added
  when a real split is requested now includes the "unsplit" readings
  from regions which apparently contained no SNPs (and therefore
  belong to all split contigs).

* Report Mutations bug fix. Producing HTML output now works (albeit
  with less content) when asking for SNPs to be derived from
  "differences" instead of tags.

Trev
----

* Pyrosequencing data (eg 454 flowgrams) are now supported. Trev
  auto-senses the type and views accordingly. Added format for 454's
  SFF standard too.

Io_lib
------

* New release - 1.9.0. No longer uses libtool in configuration.

* Added "HASH=hashfile" component of RAWDATA search path along with
  hash_tar and hash_extract tools. These operate like index_tar but
  provide true fast random access into tar files (or other formats if
  needed).

* Major update of I/O mechanisms to use in-memory equivalents of the
  files being read from or written to. This has lead to some
  incompatibilities (see src/io_lib/CHANGES), but is also an order of
  magnitude faster in many cases.

Spin
----

* Under MS Windows spin and EMBOSSwin should coexist happily to the
  extent that Spin will spot the EMBOSSwin installation and use it.

Pregap4
-------

* Fixed various space-in-pathname bugs (most prevalent on MS Windows).


Misc
----

* Removed some redundant old ABI utilities: getABIfloat, getABIhex,
  getABIraw, getABIstring. All of these can be achieved via getABIfield.


-----------------------------------------------------------------------------

3 May 2005
==========

Internal Sanger test version rel-1-6-0b3 ("gap4 -test")

Gap4
----

* Substantial modifications to the gap4 I/O mechanisms (balancing the
  "freetree"). The upshot of this is that the worst-case performance
  is now dramatically improved and the average performance will be
  slightly better too. Complementing a 10Mb contig dropped from about
  50 minutes to 1.5 minutes. Disassembling all readings (approx
  175000) in this database went from ~3 hours to 2.5 minutes.

* The contig editor "sequence sets" are now coloured to aid
  distinguishing.

* Fixed bugs with X11 BadDrawable crashes (I hope).

* Reverted the OLIGo tag back to using "serial=#" as the first line,
  although the additional information is still kept (albeit in a
  reduced precision format to make the tag shorter).

-----------------------------------------------------------------------------

29 March 2005
=============

Internal Sanger test version rel-1-6-0b2 (the previous "gap4 -test")

Gap4
----

* Added 64-bit filesize support, meaning that Gap4 databases can now
  be larger than 2Gb in file size. Specify "-bitsize 64" to force
  64-bit mode, otherwise databases will still be created in the old
  32-bit format. Copy_db has a -b flag to change bitsize when copying
  databases so this aids in migration.

* A new experimental option "SNP candidates" (in the main view
  menu). Primarily designed for pulling apart the two alleles from
  mixed shotguns, it also has uses for exploring and sometimes
  resolving collapsed repeats. User feedback at this stage is most
  welcome.

* The consistency plots now distinguish between the old "discrepancy
  plot" (now named diploid discrepancy plot) and a new discrepancy
  plot that takes does not have the "two alleles only in 50/50 ratios"
  detector. This makes it more suitable for finding potential
  collapsed repeats.

* The contig editor now has the notion of "sets". A set is basically a
  sub-group of a contig, which is useful for exploring the outcome
  within a conti of pulling apart a collapsed repeat without actually
  doing it. The editor is automatically invoked in this manner by the
  SNP candidates output. This is largely experimental, but there are a
  number of potential ways to improve interaction. Ideas on a postcard
  please!

* Lots of speedups (primarily for the huge 64-bit databases):
  - Drawing of some consistency graphs;
  - Experiment file reading on large files (1000 fold on a 1Mb sequence);
  - Dealing with lists of contigs (eg "all contigs") is now O(N)

src/init_exp/init_exp.c
/*
 * Title: init_exp
 *
 * Based upon getMCH.c
 */

#include <stdio.h>
#include <string.h>

#include "Read.h"
#include "expFileIO.h"
#include "errno.h"
#include "misc.h"
#include "open_trace_file.h"

double avg_qual(Read *r) {
    double aq = 0.0;
    int i;
    int i_start = 100, i_end = 200;

    if (i_start >= r->NBases)
        return 0.0;

    if (i_end > r->NBases)
        i_end = r->NBases;

    for (i = i_start ; i < i_end; i++) {
        switch(r->base[i]) {
        case 'a':
        case 'A':
            aq += r->prob_A[i];
            break;

        case 'c':
        case 'C':
            aq += r->prob_C[i];
            break;

        case 'g':
        case 'G':
            aq += r->prob_G[i];
            break;

        case 't':
        case 'T':
            aq += r->prob_T[i];
            break;
        }
    }

    aq /= (i_end - i_start);

    return aq;
}

int convert(char *file, int format, FILE *ofp, char *name, int output_conf) {
    Read *r;
    Exp_info *e;
    char buf[50];
    double aq;

    if (format == TT_BIO) {
        if (NULL == (r = read_reading(file, format))) {
            fprintf(stderr, "%s: failed to read\n", file);
            return 1;
        }
    } else {
        FILE *infp;
        if (NULL == (infp = open_trace_file(file, NULL))) {
            perror(file);
            return 1;
        }
        if (NULL == (r = fread_reading(infp, file, format))) {
            fprintf(stderr, "%s: failed to read\n", file);
            return 1;
        }
        fclose(infp);
    }

    e = read2exp(r, name);
    if (NULL == e) {
        fprintf(stderr, "Failed to create experiment file.\n");
        read_deallocate(r);
        return 1;
    }

    sprintf(buf, "%f", aq = avg_qual(r));
    exp_set_entry(e, EFLT_AQ, buf);
    exp_print_file(ofp, e);

    if (output_conf && aq != 0) {
        char *cstr;
        int1 *conf;
        int i;

        conf = xmalloc(r->NBases * sizeof(*conf));
        cstr = xmalloc(5 * r->NBases+2);
        for (i = 0; i < r->NBases; i++) {
            switch (r->base[i]) {
            case 'a':
            case 'A':
                conf[i] = r->prob_A[i];
                break;
            case 'c':
            case 'C':
                conf[i] = r->prob_C[i];
                break;
            case 'g':
            case 'G':
                conf[i] = r->prob_G[i];
                break;
            case 't':
            case 'T':
                conf[i] = r->prob_T[i];
                break;
            default:
                conf[i] = (r->prob_A[i] +
                           r->prob_C[i] +
                           r->prob_G[i] +
                           r->prob_T[i]) / 4;
                break;
            }
        }

        conf2str(conf, r->NBases, cstr);
        exp_set_entry(e, EFLT_AV, cstr);

        xfree(cstr);
        xfree(conf);
    }

    read_deallocate(r);
    exp_destroy_info(e);

    return 0;
}


void usage(void) {
    fprintf(stderr,
            "usage: init_exp [-(abi|alf|scf|pln)] [-output file] "
            "[-name entry_name] [-conf] file\n");
    exit(1);
}

int main(int argc, char **argv) {
    int a = 1;
    int format = TT_ANY;
    FILE *ofp = stdout;
    char *name = NULL, *oname = NULL;
    int output_conf = 0;

    read_sections(READ_BASES);

    while (a < argc) {
        if (strcasecmp(argv[a], "-abi") == 0)
            format = TT_ABI;

        else if (strcasecmp(argv[a], "-alf") == 0)
            format = TT_ALF;

        else if (strcasecmp(argv[a], "-scf") == 0)
            format = TT_SCF;

        else if (strcasecmp(argv[a], "-pln") == 0)
            format = TT_PLN;

        else if (strcasecmp(argv[a], "-name") == 0) {
            if (a == argc)
                usage();

            name = argv[++a];

        } else if (strcasecmp(argv[a], "-output") == 0) {
            if (a == argc)
                usage();

            ofp = fopen(oname = argv[++a], "wb");
            if (NULL == ofp) {
                perror(argv[a]);
                return 1;
            }

        } else if (strcasecmp(argv[a], "-conf") == 0) {
            output_conf = 1;

        } else if (argv[a][0] == '-')
            usage();
        else
            break;

        a++;
    }

    if (a+1 != argc)
        usage();

    if (!name)
        name = oname;
    if (!name)
        name = argv[a];

    return convert(argv[a], format, ofp, name, output_conf);
}
